databricks/koalas/frame.py

```diff
@@ -2042,6 +2042,248 @@ class  locomotion
+
+        >>> df1.where(cond).sort_index()
+             A      B
+        0  NaN    NaN
+        1  1.0  200.0
+        2  NaN    NaN
+        3  NaN    NaN
+        4  NaN    NaN
+
+        >>> reset_option("compute.ops_on_diff_frames")
+        """
+        from databricks.koalas.series import Series
+        tmp_cond_col_name = '__tmp_cond_col_{}__'
+        tmp_other_col_name = '__tmp_other_col_{}__'
+        kdf = self.copy()
+        if isinstance(cond, DataFrame):
+            for column in self._internal.data_columns:
+                kdf[tmp_cond_col_name.format(column)] = cond.get(column, False)
+                if isinstance(other, DataFrame):
+                    kdf[tmp_other_col_name.format(column)] = other.get(column, np.nan)
+                else:
+                    kdf[tmp_other_col_name.format(column)] = other
+        elif isinstance(cond, Series):
+            for column in self._internal.data_columns:
+                kdf[tmp_cond_col_name.format(column)] = cond
+                if isinstance(other, DataFrame):
+                    kdf[tmp_other_col_name.format(column)] = other.get(column, np.nan)
+                else:
+                    kdf[tmp_other_col_name.format(column)] = other
+        else:
+            raise ValueError("type of cond must be a DataFrame or Series")
+
+        sdf = kdf._sdf
+        # above logic make spark dataframe looks like below:
+        # +-----------------+---+---+------------------+-------------------+------------------+--...
+        # |__index_level_0__|  A|  B|__tmp_cond_col_A__|__tmp_other_col_A__|__tmp_cond_col_B__|__...
+        # +-----------------+---+---+------------------+-------------------+------------------+--...
+        # |                0|  0|100|              true|                  0|             false|  ...
+        # |                1|  1|200|             false|                 -1|             false|  ...
+        # |                3|  3|400|              true|                 -3|             false|  ...
+        # |                2|  2|300|             false|                 -2|              true|  ...
+        # |                4|  4|500|             false|                 -4|             false|  ...
+        # +-----------------+---+---+------------------+-------------------+------------------+--...
+
+        output = []
+        for column in self._internal.data_columns:
+            data_col_name = self._internal.column_name_for(column)
+            output.append(
+                F.when(
+                    sdf[tmp_cond_col_name.format(column)], sdf[data_col_name]
+                ).otherwise(
+                    sdf[tmp_other_col_name.format(column)]
+                ).alias(data_col_name))
+
+        index_columns = self._internal.index_columns
+        sdf = sdf.select(*index_columns, *output)
+        internal = self._internal.copy(sdf=sdf)
+
+        return DataFrame(_InternalFrame(sdf=sdf, index_map=self._internal.index_map))
```

Comment:
Hm, why can't we `self._internal.copy(sdf=sdf)`?

Reply:
yeah, i think we can just use copy here with `sdf` and `column_scols`  since i think maybe `self._internal.sdf` and `sdf` are not same identically.